target_decisioning_engine/tests/test_artifact_provider.py

```diff
@@ -0,0 +1,257 @@
+        self.provider.add_subscription(subscriber)
+        self.assertEqual(self.provider.subscription_count, 1)
+        self.assertEqual(self.provider.subscriptions[1], subscriber)
+
+        # Test remove_subscription
+        self.provider.remove_subscription(1)
+        self.assertEqual(self.provider.subscription_count, 0)
+
+    def test_remove_subscription_bad_key(self):
+        self.provider = ArtifactProvider(self.default_config)
+        self.assertEqual(self.provider.subscription_count, 0)
+        self.provider.remove_subscription(100)
+        self.assertEqual(self.provider.subscription_count, 0)
+
+    def test_emit_new_artifact(self):
+        emitter = Mock()
+        subscriber = Mock()
+        config = DecisioningConfig("client123", "org999", event_emitter=emitter)
+        self.provider = ArtifactProvider(config)
+        self.provider.add_subscription(subscriber)
+        self.assertEqual(self.provider.subscription_count, 1)
+
+        payload = {"a": 1, "b": 2}
+        self.provider.emit_new_artifact(payload)
+
+        call1 = call(ARTIFACT_DOWNLOAD_SUCCEEDED, {
+            "artifact_location": self.provider.artifact_location,
+            "artifact_payload": payload
+        })
+        call2 = call(GEO_LOCATION_UPDATED, {
+            "geo_context": {}
+        })
+
+        emitter.assert_has_calls([call1, call2], any_order=False)
+        subscriber.assert_called_with(payload)
+
+    def test_stop_all_polling(self):
+        with patch.object(ArtifactProvider, "fetch_artifact", return_value=None):
+            self.provider = ArtifactProvider(self.default_config)
+            self.provider.initialize()  # starts polling thread
+            self.assertTrue(self.provider.polling_timer.is_alive())
+
+            self.provider.stop_all_polling()
+            self.assertIsNone(self.provider.polling_timer)
+            self.assertTrue(self.provider.polling_halted)
+
+    def test_resume_polling(self):
+        with patch.object(ArtifactProvider, "fetch_artifact", return_value=None):
+            self.provider = ArtifactProvider(self.default_config)
+            self.provider.initialize()  # starts polling thread
+            self.provider.stop_all_polling()
+            self.assertTrue(self.provider.polling_halted)
+
+            self.provider.resume_polling()
+            self.assertTrue(self.provider.polling_timer.is_alive())
+            self.assertFalse(self.provider.polling_halted)
+
+    def test_schedule_next_update_polling_halted(self):
+        with patch.object(ArtifactProvider, "fetch_artifact", return_value=None):
+            self.provider = ArtifactProvider(self.default_config)
+            self.provider.initialize()  # starts polling thread
+            self.assertTrue(self.provider.polling_timer.is_alive())
+            self.assertFalse(self.provider.polling_halted)
+
+            self.provider.stop_all_polling()
+            self.provider.schedule_next_update()
```

Comment:
Same thing here. `schedule_next_update` should not be called by your test.  It needs to be called internal to ArtifactProvider.  I don't remember if python supports private methods, but you should not even be able to call `schedule_next_update` outside of ArtifactProvider.